### PR TITLE
Make `TagCheck` own `Predicate` and remove predicate arena.

### PR DESCRIPTION
**Pull request by bgogul**
This is in preparation for implementing the core data structures. Now that we do not need to instantiate the access paths for printing datalog facts, we don't need this predicate arena trick. This PR reverts most of the changes from PR #206.

### CODE REVIEW

File: src/ir/predicate.cc [hunk 1]
```diff
@@ -15,35 +15,32 @@
...
+                               std::unique_ptr<Predicate> rhs) {
+  return std::unique_ptr<Or>(new Or(std::move(lhs), std::move(rhs)));
 }
 
-const TagPresence *TagPresence::Create(PredicateArena &arena, std::string tag) {
-  return arena.CapturePredicate(std::unique_ptr<TagPresence>(
-      new TagPresence(std::move(tag))));
+std::unique_ptr<TagPresence> TagPresence::Create(std::string tag) {
+  return std::unique_ptr<TagPresence>(new TagPresence(std::move(tag)));
 }
```
[the remark this answers is not in the document]
Comment (bgogul) [reply src/ir/predicate.cc:44]: I meant to move them after getting the tests fixed. Will do so.

File: src/ir/predicate_test.cc [hunk 2]
```diff
@@ -90,125 +90,75 @@ INSTANTIATE_TEST_SUITE_P(ToDatalogRuleBodyTest, ToDatalogRuleBodyTest,
...
+
+// For this test, we create a number of expected-distinct predicates and try
+// all pairs of them. We then ensure that all of these predicates are equal
+// only when their pointers are equal.
 TEST_P(PredicateEqTest, PredicateEqTest) {
-  EXPECT_EQ(*pred1_ == *pred2_, pred_num1 == pred_num2);
+  const Predicate *pred1 = std::get<0>(GetParam());
+  const Predicate *pred2 = std::get<1>(GetParam());
+
+  EXPECT_EQ(pred1 == pred2, *pred1 == *pred2);
```
[the remark this answers is not in the document]
Comment (bgogul) [reply src/ir/predicate_test.cc:104]: Ah, I thought you only needed that when you moved to predicate arena in #206. I went back to the conversation and I think I misunderstood what you meant. Will put these back in.

File: src/ir/predicate.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (bgogul) [reply src/ir/predicate.cc]: This is done. I removed the factory methods as well.

File: src/ir/predicate_test.cc [hunk 3]
```diff
@@ -116,94 +116,78 @@ TEST_P(PredicateEqTest, PredicateEqTest) {
...
-    { kNotTag1.predicate(), 12},
-    { kNotTag1_2.predicate(), 12},
-    { kNotTag2.predicate(), 13},
-    { kNotTag2_2.predicate(), 13},
-    { kNotTag3.predicate(), 14},
-    { kNotTag3_2.predicate(), 14}
-};
+    {&kTag1Present, 0},      {&kTag1Present2, 0},
+    {&kTag2Present, 1},      {&kTag2Present2, 1},
+    {&kTag3Present, 2},      {&kTag3Present2, 2},
```
Comment (markww) [inline src/ir/predicate_test.cc:172]: I think this formatting is really readable and nice!

**Review by markww — APPROVED** [review]
LGTM

File: src/ir/predicate_test.cc (hunk 3 above)
Comment (bgogul) [reply src/ir/predicate_test.cc:172]: Well, `clang-format` did it for me. :)

**Pull request merged**